File: modelica/test/slice1_fuelpin_steady.py
"""Slice 1 — single steady fuel pin, Modelica acausal network vs textbook closed form.

PURPOSE (verification, not validation): confirm Modelica's connect() assembles
and solves the across/through thermal network correctly, by checking the four
node temperatures against the independently-computed analytic series-resistance
result to machine precision. Constant properties on purpose — this isolates the
SOLVER MACHINERY from the closures. Matching netflow's T-dependent number is
slice 4. See docs/adr/0001.

Standing limits (greppable, per Soul completion gate):
DEBT: constant properties (k_fuel, k_clad, h_gap) — verification only.
      T-dependent UO2/He/Zr closures + Dittus-Boelter convection come in slice 4.
"""
from __future__ import annotations

import math
import sys

import DyMat

from dymola_harness import new_dymola, simulate


# Geometry / operating point (must match Components.FuelPinConst defaults)
R_PELLET = 4.10e-3
GAP_THICK = 0.085e-3
R_CO = 4.75e-3
L = 1.0
Q_LIN = 18_000.0
K_FUEL = 3.0
K_CLAD = 16.0
H_GAP = 5_000.0
T_COOL = 593.0


def analytic_node_temperatures() -> dict[str, float]:
    """Independently re-derived from the same geometry & properties.

    Series resistances [K/W], pellet uses the solid-cylinder uniform-generation
    closed form R = 1 / (4 pi k L).
    """
    r_ci = R_PELLET + GAP_THICK
    Q_total = Q_LIN * L

    R_fuel = 1.0 / (4.0 * math.pi * K_FUEL * L)
    R_gap = 1.0 / (2.0 * math.pi * R_PELLET * H_GAP * L)
    R_clad = math.log(R_CO / r_ci) / (2.0 * math.pi * K_CLAD * L)

    T_co = T_COOL  # pinned by FixedTemperature; no convection resistance in slice 1
    T_ci = T_co + Q_total * R_clad
    T_ps = T_ci + Q_total * R_gap
    T_cl = T_ps + Q_total * R_fuel
    return {
        "centerline": T_cl,
        "pellet_surface": T_ps,
        "clad_inner": T_ci,
        "clad_outer": T_co,
    }


def read_modelica_steady(mat_path: str) -> dict[str, float]:
    m = DyMat.DyMatFile(mat_path)
    # Trajectories are constant (no states); take the final point.
    return {
        "centerline": float(m.data("pin.T_centerline_K")[-1]),
        "pellet_surface": float(m.data("pin.T_pellet_surface_K")[-1]),
        "clad_inner": float(m.data("pin.T_clad_inner_K")[-1]),
        "clad_outer": float(m.data("pin.T_clad_outer_K")[-1]),
    }


def main() -> int:
    anal = analytic_node_temperatures()
    print("analytic (independently re-derived):")
    for k, v in anal.items():
        print(f"  {k:<16s} = {v:.6f} K")

    with new_dymola() as (dymola, work):
        mat = simulate(
            dymola, work,
            "NetflowModelica.Tests.FuelPinSteady",
            stop_time=1.0,
            n_intervals=10,
            result_stem="slice1",
        )
        mod = read_modelica_steady(mat)

    print("\nmodel:")
    for k, v in mod.items():
        print(f"  {k:<16s} = {v:.6f} K")

    print("\n node              model [K]    analytic [K]   |Δ| [K]")
    max_d = 0.0
    for k in anal:
        d = abs(mod[k] - anal[k])
        max_d = max(max_d, d)
        print(f"  {k:<16s} {mod[k]:11.4f}  {anal[k]:13.4f}   {d:.2e}")

    print(f"\n max node Δ = {max_d:.2e} K")
    # Machine-precision target — Julia's slice 1 hit ~0 (linear system; one Newton step).
    # Loosened slightly to account for Dymola's translate-time numerical evaluation
    # of the resistance parameters in single-precision-promoted intermediates.
    threshold = 1e-9
    if max_d < threshold:
        print(f"VERIFY PASS (machine-precision agreement, Δ < {threshold:g} K)")
        return 0
    else:
        print(f"VERIFY FAIL (Δ >= {threshold:g} K)")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: julia-mtk/test/slice1_fuelpin_steady.jl
# Slice 1 — single steady fuel pin, MTK acausal network vs textbook closed form.
#
# PURPOSE (verification, not validation): confirm ModelingToolkit's connect()
# assembles and solves the across/through thermal network correctly, by checking
# the centerline temperature against the independently-computed analytic
# series-resistance result to machine precision. Constant properties on purpose —
# this isolates the SOLVER MACHINERY from the closures. Matching netflow's
# T-dependent number is a later slice. See docs/adr/0001.
#
# Standing limits (greppable, per Soul completion gate):
# DEBT: constant properties (k_fuel, k_clad, h_gap, h_conv) — machinery verification
#       only; T-dependent UO2 conductivity + netflow's gap/convection closures are
#       needed to match netflow's 1201.9 K (slice 3+). The 1252.99 K here is NOT that.
# TODO: NonlinearSolve retcode reads `Stalled` (linear system already at machine-zero
#       residual on step 1); a LinearProblem/SteadyStateProblem would report `Success`.
#       Correctness is established by the analytic comparison, which is retcode-independent.

using ModelingToolkit
using ModelingToolkitStandardLibrary.Thermal
using NonlinearSolve
using ModelingToolkit: t_nounits as t

# ---- Geometry (netflow's documented pin) ------------------------------------
r_pellet   = 4.10e-3            # m
gap_thick  = 0.085e-3           # m
r_ci       = r_pellet + gap_thick   # clad inner radius, m
r_co       = 4.75e-3            # clad outer radius, m
L          = 1.0               # m
q_lin      = 18_000.0          # W/m
Qtot       = q_lin * L         # W (total power into the pin)

# ---- Representative CONSTANT properties (verification anchor) ----------------
k_fuel = 3.0      # W/m-K  (constant; real UO2 is T-dependent — later)
k_clad = 16.0     # W/m-K
h_gap  = 5_000.0  # W/m^2-K, referenced at r_pellet
h_conv = 30_000.0 # W/m^2-K
T_cool = 593.0    # K

# ---- Analytic series resistances [K/W] --------------------------------------
R_fuel = 1 / (4π * k_fuel * L)
R_gap  = 1 / (2π * r_pellet * h_gap * L)
R_clad = log(r_co / r_ci) / (2π * k_clad * L)
R_conv = 1 / (2π * r_co * h_conv * L)
R_tot  = R_fuel + R_gap + R_clad + R_conv

T_center_analytic = T_cool + Qtot * R_tot
println("R_tot = $R_tot K/W ; analytic centerline = $T_center_analytic K")

# ---- MTK acausal network ----------------------------------------------------
@named src  = FixedHeatFlow(Q_flow = Qtot)   # positive Q_flow injects heat into network
@named Rf   = ThermalResistor(R = R_fuel)
@named Rg   = ThermalResistor(R = R_gap)
@named Rc   = ThermalResistor(R = R_clad)
@named Rh   = ThermalResistor(R = R_conv)
@named cool = FixedTemperature(T = T_cool)

eqs = [
    connect(src.port, Rf.port_a),
    connect(Rf.port_b, Rg.port_a),
    connect(Rg.port_b, Rc.port_a),
    connect(Rc.port_b, Rh.port_a),
    connect(Rh.port_b, cool.port),
]

@named pin = System(eqs, t; systems = [src, Rf, Rg, Rc, Rh, cool])
sys = mtkcompile(pin)

println("unknowns: ", unknowns(sys))

prob = NonlinearProblem(sys, [src.port.T => T_cool])
sol  = solve(prob, NewtonRaphson())
println("retcode: ", sol.retcode)

T_center_mtk = sol[src.port.T]
err = abs(T_center_mtk - T_center_analytic)
println("centerline MTK     = $T_center_mtk K")
println("centerline analytic= $T_center_analytic K")
println("abs error          = $err K")
println(err < 1e-6 ? "VERIFY PASS (machine-precision agreement)" : "VERIFY FAIL")
